### PR TITLE
feature(azure-iot-device): Support for Commands using Digital Twin Mode

### PR DESCRIPTION
**Pull request by cartertinney**
* Added `.send_command_response()` method to both Device and Module clients
* Added `.on_command_received` handler to both Device and Module clients
* Refactored testing infrastructure in order to support more rigorous handler testing.

### CODE REVIEW

File: azure-iot-device/azure/iot/device/iothub/abstract_clients.py [hunk 1]
```diff
@@ -507,7 +506,10 @@ def on_method_request_received(self):
         """The handler function or coroutine that will be called when a method request is received.
 
         The function or coroutine definition should take one positional argument (the
-        :class:`azure.iot.device.MethodRequest` object)"""
+        :class:`azure.iot.device.MethodRequest` object)
+
+        This handler is not compatible with PNP.
```
Comment (cartertinney) [inline azure-iot-device/azure/iot/device/iothub/abstract_clients.py:511]: I know we probably don't want to put PNP anywhere publicly. Should I refer to this as "Digital Twin" instead perhaps?

**Comment by cartertinney** [comment]
Apologies for there being some confusing merge from master stuff in this PR - some of these changes are actually just other stuff that was in the master branch that got wrapped into here. I can try to resubmit this PR without some extraneous changes if we don't want the changes from master.

File: azure-iot-device/azure/iot/device/iothub/aio/async_clients.py [hunk 2]
```diff
@@ -545,6 +544,44 @@ async def receive_twin_desired_properties_patch(self):
         logger.info("twin patch received")
         return patch
 
+    async def send_command_response(self, command_response):
+        """Send a response to a command via the Azure IoT Hub or Azure IoT Edge Hub.
+
+        If the connection to the service has not previously been opened by a call to connect, this
+        function will open the connection before sending the event.
```
Comment (BertKleewein) [inline azure-iot-device/azure/iot/device/iothub/aio/async_clients.py:551]: >event [](http://example.com/codeflow?start=61&length=5)

*response

File: azure-iot-device/azure/iot/device/iothub/aio/async_clients.py [hunk 3]
```diff
@@ -545,6 +544,44 @@ async def receive_twin_desired_properties_patch(self):
...
+
+        This method is only compatible with PNP.
```
Comment (BertKleewein) [inline azure-iot-device/azure/iot/device/iothub/aio/async_clients.py:553]: >PNP [](http://example.com/codeflow?start=44&length=3)

*Azure IoT Digital Twins
here and everywhere

**Review by BertKleewein — APPROVED** [review]
:shipit:

**Comment by cartertinney** [comment]
Closing due to pipeline issue, this is now tracked in #815

**Pull request closed without merging**